File: 2024/day10.cpp
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>

#include "../util.hpp"

int main(int, char**){
    std::vector<std::string> hikemap = util::splitline(util::getInput(2024, 10));
    int n = hikemap.size(), m = hikemap[0].size();

    // part 1
    std::unordered_map<int, std::unordered_set<int>> buff;
    for (int i = 0; i < n; ++i){
        for (int j = 0; j < m; ++j){
            if (hikemap[i][j] == '9'){
                buff[i * m + j].insert(i * m + j);
            }
        }
    }
    std::unordered_map<int, std::unordered_set<int>> newBuff;
    for (int k = 0; k < 9; ++k){
        newBuff.clear();
        for (auto &pr : buff){
            int i = pr.first / m, j = pr.first % m;
            for (const auto &nb : util::neighbourGridPoint(i, j, n, m)){
                if (hikemap[nb.first][nb.second] == hikemap[i][j] - 1){
                    newBuff[nb.first * m + nb.second].insert(pr.second.cbegin(), pr.second.cend());
                }
            }
        }
        std::swap(buff, newBuff);
    }
    int res = 0;
    for (auto &pr : buff){
        res += pr.second.size();
    }
    util::output(res);

    // part 2
    std::unordered_map<int, int> buffPart2;
    for (int i = 0; i < n; ++i){
        for (int j = 0; j < m; ++j){
            if (hikemap[i][j] == '9'){
                buffPart2[i * m + j] = 1;
            }
        }
    }
    std::unordered_map<int, int> newBuffPart2;
    for (int k = 0; k < 9; ++k){
        newBuffPart2.clear();
        for (const auto &pr : buffPart2){
            int i = pr.first / m, j = pr.first % m;
            for (const auto &nb : util::neighbourGridPoint(i, j, n, m)){
                if (hikemap[nb.first][nb.second] == hikemap[i][j] - 1){
                    newBuffPart2[nb.first * m + nb.second] += pr.second;
                }
            }
        }
        std::swap(buffPart2, newBuffPart2);
    }
    res = 0;
    for (const auto &pr : buffPart2){
        res += pr.second;
    }
    util::output(res);
    return 0;
}

File: 2024/day10.py
import AOCInit
import util

from collections import defaultdict

if __name__ != '__main__':
    exit()

inp = util.getInput(d=10, y=2024)

hikemap = inp.splitlines()
n = len(hikemap)
m = len(hikemap[0])

timer = util.Timer()

# part 1
buff = defaultdict(set)
for i in range(n):
    for j in range(m):
        if hikemap[i][j] == '9':
            buff[(i, j)].add(i * m + j)
for _ in range(9):
    newBuff: defaultdict[tuple[int, int], set[int]] = defaultdict(set)
    for pt, indices in buff.items():
        for ii, jj in util.nearby2DGridPts(pt, (n, m)):
            if ord(hikemap[ii][jj]) == ord(hikemap[pt[0]][pt[1]]) - 1:
                newBuff[(ii, jj)] |= indices
    buff = newBuff
print(sum(len(v) for v in buff.values()))

timer.check()

# part 2
buffPart2 = defaultdict(int)
for i in range(n):
    for j in range(m):
        if hikemap[i][j] == '9':
            buffPart2[(i, j)] = 1
for _ in range(9):
    newBuffPart2: defaultdict[tuple[int, int], int] = defaultdict(int)
    for pt, val in buffPart2.items():
        for ii, jj in util.nearby2DGridPts(pt, (n, m)):
            if ord(hikemap[ii][jj]) == ord(hikemap[pt[0]][pt[1]]) - 1:
                newBuffPart2[(ii, jj)] += val
    buffPart2 = newBuffPart2
print(sum(buffPart2.values()))

timer.stop()

